man/schleuder.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SCHLEUDER" "8" "January 2017" "" ""
.
.SH "NAME"
\fBschleuder\fR \- an email hub for groups
.
.SH "SYNOPSIS"
\fBschleuder\fR work \fIlistaddress\fR < \fIemail\fR
.
.P
\fBschleuder\fR help
.
.P
\fBschleuder\fR \fIother\-command\fR [<\.\.\.>]
.
.P
For descriptions of the other commands see the output of the \fIhelp\fR\-command\.
.
.P
This manual page is written for system administrators\. For other information please read the Schleuder documentation online (\fISEE ALSO\fR)\.
.
.SH "DESCRIPTION"
Schleuder is an email hub for groups\. Subscribers can communicate encryptedly and pseudonymously among themselves, receive emails from non\-subscribers and send emails to non\-subscribers via the list\.
.
.P
Schleuder takes care of all de\- and encryption, stripping of headers, formatting conversions, etc\. Further Schleuder can send out its own public key and receive administrative commands by email\.
.
.P
Email cryptography is handled by using GnuPG\.
.
.SH "ENVIRONMENT"
.
.SS "Configuration"
Schleuder reads its basic settings from a file that it by default expects at </etc/schleuder/schleuder\.yml>\. To make Schleuder read a different file set the environment variable \fISCHLEUDER_CONFIG\fR to the path to your file when running schleuder\. E\.g\.:
.
.IP "" 4
.
.nf

SCHLEUDER_CONFIG=/usr/local/etc/schleuder\.yml /path/to/bin/schleuder \.\.\.
.
.fi
.
.IP "" 0
.
.P
For explanations of the possible settings read the default config file\.
.
.P
The default settings for new lists are read from another config file\. By default Schleuder looks at </etc/schleuder/list\-defaults\.yml>\. To make Schleuder read a different file set the environment variable \fISCHLEUDER_LIST_DEFAULTS\fR analogous to above\. The possible settings are explained in the default config file\.
.
.SS "Connect to MTA"
Schleuder behaves like an email\-filter: it reads email from standard\-input, and reports errors to standard\-error\. If all goes well Schleuder closes the initial connection to the Mail Transport Agent (MTA) only after it sent out all outgoing emails\.
.
.P
In case of an error the MTA is expected to include Schleuder\'s error message into a bounce\-email that is sent back to the sender (this is default behaviour of most MTAs)\.
.
.P
To connect the MTA with Schleuder it must pipe the incoming message into Schleuder\'s STDIN, and give it two arguments: first one: "work", second one: the full recipients email address\.
.
.P
For more information on how to integrate Schleuder with your existing mail setup, please read the Schleuder documentation online (\fISEE ALSO\fR)\.
.
.SS "Data storage"
The keyrings for each list are standard GnuPG keyrings and sit in the filesystem under \fIlists_dir\fR/\fIhostname\fR/\fIlistname\fR/ (\fIlists_dir\fR is read from schleuder\.yml, by default it is </var/lib/schleuder/lists>)\. They can be used manually using gpg2\. Please be careful to maintain proper file permissions if you touch the files\.
.
.P
In the list\-directory there’s also a list specific log\-file (might be missing if the log\-level is high and no error occurred yet)\.
.
.P
Other logging is sent to syslog\. Where that ends up depends on the operating system and the system administration\.
.
.P
All other list\-related data is stored in the SQL\-database\. Most data is unserialized, only some values are JSON\-encoded\.
.
.SH "SPECIAL FEATURES"
Schleuder features some special functionality\. For more detailed information read the Schleuder documentation online (\fISEE ALSO\fR)\.
.
.SS "Getting the public key of a list"
Each Schleuder\-list replies with its public key to any email sent to \fIlistname\-sendkey@hostname\fR\. E\.g\. to receive the key for the contact address of the Schleuder project write an email to
.
.br
\fIteam\-sendkey@schleuder\.org\fR\.
.
.SS "Email commands"
Schleuder knows some special keywords that trigger different behaviour\. You can e\.g\. subscribe someone, or resend an email to a non\-subscriber using keywords\.
.
.P
Keywords require that:
.
.IP "\(bu" 4
they start the line and begin with "x\-",
.
.IP "\(bu" 4
they are written into the beginning of the \fIfirst text\-part\fR of the email (usually that’s just the normal body of the email),
.
.IP "\(bu" 4
possible arguments must be written \fIon the same line\fR as the keyword (exceptions are mentioned in the descriptions below),
.
.IP "\(bu" 4
the email must be \fIencrypted and signed\fR by a list\-member’s key\.
.
.IP "" 0
.
.P
Keywords can be repeated within one email at will\. Letter case doesn’t matter\.
.
.P
There are two types of keywords: those to enhance messages sent over the list (“list\-keywords”), and those to request something from Schleuder (“request\-keywords”)\.
.
.P
Find detailed descriptions of all available keywords in the Schleuder documentation online (\fISEE ALSO\fR)\.
.
.SS "Contact list\-owner"
Write to \fIlistname\-owner@hostname\fR to contact the list\-owner(s) even if you don\'t know who they are\. Use the list\'s key to encrypt the email!
.
.SH "EXIT STATUS"
.
.TP
0
Incoming email was processed without errors\.
.
.TP
1
Internal failure in incoming email processing\.
.
.SH "FILES"
.
.IP "\(bu" 4
\fB/etc/schleuder/schleuder\.yml\fR: default path of global Schleuder configuration
.
.IP "\(bu" 4
\fB/etc/schleuder/list\-defaults\.yml\fR: default path of default list settings
.
.IP "\(bu" 4
\fB/var/lib/schleuder/lists\fR default path of lists_dir
.
.IP "\(bu" 4
\fB<lists_dir>\fR/\fB<hostname>\fR/`\fIlistname\fR: list internal data
.
.IP "\(bu" 4
\fB<lists_dir>\fR/\fB<hostname>\fR/\fB<listname>\fR/list\.log`: log\-file for list
.
.IP "" 0
.
.P
All configuration files are formatted as YAML\. See \fIhttp://www\.yaml\.org/\fR for more details\.
.
.SH "BUGS"
Known bugs are listed on the Schleuder bugtracker at \fIhttps://0xacab\.org/schleuder/schleuder\fR
.
.SH "SEE ALSO"
\fBschleuder\-cli\fR(8), \fBgnupg\fR(7)\.
.
.TP
Website of \fBschleuder\fR
\fIhttps://schleuder\.org/\fR
.
.TP
More extensive documentation for \fBschleuder\fR
\fIhttps://schleuder\.org/docs/\fR
.
.TP
\fBschleuder\-cli\fR, the command line interface for list\-management
\fIhttps://0xacab\.org/schleuder/schleuder\-cli/\fR
.
.TP
\fBschleuder\-web\fR, the web interface for list\-management
\fIhttps://0xacab\.org/schleuder/schleuder\-web/\fR



NEWS
schleuder (3.2.1-1) unstable; urgency=medium

  Schleuder is now able to use the pre-configured sensible keyserver defaults
  of dirmngr.
  To make use of this feature, specify a blank value in /etc/schleuder.yml:
  keyserver:

 -- Georg Faerber <georg@riseup.net>  Mon, 09 Oct 2017 20:22:19 +0200

schleuder (3.1.2-1) unstable; urgency=medium

  In case schleuder was upgraded from a version < 3.0, all existing list data
  was moved to '/var/backup/schleuder'. 
  Please see '/usr/share/doc/schleuder/README.Debian' for details on how to
  migrate this data to be used with the current version.

 -- Georg Faerber <georg@riseup.net>  Thu, 13 Jul 2017 13:20:13 +0200


lib/schleuder.rb
# Stdlib
require 'fileutils'
require 'singleton'
require 'yaml'
require 'pathname'
require 'syslog/logger'
require 'logger'
require 'open3'

# Require mandatory libs. The database-layer-lib is required below.
require 'mail-gpg'
require 'active_record'

# An extra from mail-gpg
require 'hkp'

# Load schleuder
libdir = Pathname.new(__FILE__).dirname.realpath
rootdir = libdir.dirname
$:.unshift libdir

# Monkeypatches
require 'schleuder/mail/parts_list.rb'
require 'schleuder/mail/message.rb'
require 'schleuder/mail/gpg.rb'
require 'schleuder/mail/encrypted_part.rb'
require 'schleuder/gpgme/import_status.rb'
require 'schleuder/gpgme/key.rb'
require 'schleuder/gpgme/sub_key.rb'
require 'schleuder/gpgme/ctx.rb'
require 'schleuder/gpgme/user_id.rb'

# The Code[tm]
require 'schleuder/errors/base'
Dir["#{libdir}/schleuder/errors/*.rb"].each do |file|
  require file
end
# Load schleuder/conf before the other classes, it defines constants!
require 'schleuder/conf'
require 'schleuder/version'
require 'schleuder/logger_notifications'
require 'schleuder/logger'
require 'schleuder/listlogger'
require 'schleuder/plugin_runners/base'
require 'schleuder/plugin_runners/list_plugins_runner'
require 'schleuder/plugin_runners/request_plugins_runner'
Dir["#{libdir}/schleuder/plugins/*.rb"].each do |file|
  require file
end
require 'schleuder/filters_runner'
Dir["#{libdir}/schleuder/validators/*.rb"].each do |file|
  require file
end
require 'schleuder/runner'
require 'schleuder/list'
require 'schleuder/list_builder'
require 'schleuder/subscription'
require 'schleuder/account'

require 'schleuder/policies/base_policy'
require 'schleuder/policies/subscription_policy'
require 'schleuder/policies/list_policy'
require 'schleuder/policies/authorizer'

# Setup
ENV['SCHLEUDER_CONFIG'] ||= '/etc/schleuder/schleuder.yml'
ENV['SCHLEUDER_LIST_DEFAULTS'] ||= '/etc/schleuder/list-defaults.yml'
ENV['SCHLEUDER_ENV'] ||= 'production'
ENV['SCHLEUDER_ROOT'] = rootdir.to_s

GPGME::Ctx.set_gpg_path_from_env
GPGME::Ctx.check_gpg_version

# TODO: Test if database is writable if sqlite.
ActiveRecord::Base.establish_connection(Schleuder::Conf.database)
ActiveRecord::Base.logger = Schleuder.logger

Mail.defaults do
  delivery_method :smtp, Schleuder::Conf.smtp_settings.symbolize_keys
end

I18n.load_path += Dir["#{rootdir}/locales/*.yml"]
I18n.enforce_available_locales = true
I18n.default_locale = :en

File.umask(0027)

include Schleuder
